### PR TITLE
BackDoorTest: add test for deletion of Feedback Session #8842

### LINKED ISSUE #8842: BackDoorTest: add test for deletion of Feedback Session
https://github.com/TEAMMATES/teammates/blob/ed1ea9c502db8a30e64d10f3b96e14c8428a124e/src/test/java/teammates/test/cases/testdriver/BackDoorTest.java#L106-L109

Add the required tests to resolve this TODO

More discussion regarding the depth of test can happen in the created PR

### PR DESCRIPTION
**Pull request by tshradheya**
Fixes #8842

**Outline of Solution**

- Added the required test for feedback session deletion
- Checked if question and response of session also deleted

### CODE REVIEW

**Comment by tshradheya** [comment]
I thought if a session is deleted from datastore, then all data that is part of it is also deleted. Am I wrong in assuming so?

Then maybe I should see if this test is actually dependent on something else

**Comment by damithc** [comment]
> I thought if a session is deleted from datastore, then all data that is part of it is also deleted. Am I wrong in assuming so?

IIRC, this is not handled by datastore but our own code. Our datastore is designed like a bunch of tables (i.e., relational-style), not a graph.

**Comment by tshradheya** [comment]
Okay. I investigated and found that those two tests were passing luckily and hence I have removed them.

Ready for review

File: src/test/java/teammates/test/cases/testdriver/BackDoorTest.java [hunk 1]
```diff
@@ -103,8 +104,26 @@ public void testDeletion() {
         assertEquals(Const.StatusCodes.BACKDOOR_STATUS_SUCCESS, status);
         verifyAbsentInDatastore(courseNoEvals);
 
-        // ----------deleting Feedback Session entities-------------------------
-        // TODO: do proper deletion test
+    }
+
+    @Test
+    public void testFeedbackSessionDeletion() {
```
Comment (damithc) [inline src/test/java/teammates/test/cases/testdriver/BackDoorTest.java]: It's odd to have a `testDeletion()` and a `testFeedbackSessionDeletion()` method.

File: src/test/java/teammates/test/cases/testdriver/BackDoorTest.java [hunk 2]
```diff
@@ -103,8 +104,26 @@ public void testDeletion() {
...
+                feedbackSession1InCourse1.getCourseId());
+        assertEquals(Const.StatusCodes.BACKDOOR_STATUS_SUCCESS, status);
+        verifyAbsentInDatastore(feedbackSession1InCourse1);
+
+        // #Closed Session
+        FeedbackSessionAttributes closedSession = dataBundle.feedbackSessions.get("closedSession");
+        verifyPresentInDatastore(closedSession);
+        status = BackDoor.deleteFeedbackSession(closedSession.getFeedbackSessionName(), closedSession.getCourseId());
+        assertEquals(Const.StatusCodes.BACKDOOR_STATUS_SUCCESS, status);
+        verifyAbsentInDatastore(closedSession);
```
Comment (damithc) [inline src/test/java/teammates/test/cases/testdriver/BackDoorTest.java]: Is a cascade expected? If yes, the test need to verify the expected cascade happens. See course deletion test (line 95).

Comment (tshradheya) [reply src/test/java/teammates/test/cases/testdriver/BackDoorTest.java]: I think its expected. I experimented with various courses, questions, and feedback sessions, it deleted the related elements.

**Comment by tshradheya** [comment]
Prof @damithc made necessary changes

File: src/test/java/teammates/test/cases/testdriver/BackDoorTest.java [hunk 3]
```diff
@@ -71,6 +72,33 @@ public void testDeletion() {
...
+
+        // verify if response of deleted closed session are deleted
+        FeedbackResponseAttributes responseOfClosedSession =
+                dataBundle.feedbackResponses.get("response1Q1ClosedPeriodFeedback");
+        verifyAbsentInDatastore(responseOfClosedSession);
```
Comment (damithc) [inline src/test/java/teammates/test/cases/testdriver/BackDoorTest.java]: Do we need to check for questions too?

**Comment by damithc** [comment]
@Crphang can have another look? The PR has changed significantly since the previous approval.

File: src/test/java/teammates/test/cases/testdriver/BackDoorTest.java (hunk 3 above)
Comment (tshradheya) [reply src/test/java/teammates/test/cases/testdriver/BackDoorTest.java]: I checked for questions when I deleted the `session1InCourse1` so covering both responses and questions. Line 85-91

File: src/test/java/teammates/test/cases/testdriver/BackDoorTest.java [hunk 4]
```diff
@@ -71,6 +72,33 @@ public void testDeletion() {
...
+        verifyPresentInDatastore(feedbackSession1InCourse1);
+        status = BackDoor.deleteFeedbackSession(feedbackSession1InCourse1.getFeedbackSessionName(),
+                feedbackSession1InCourse1.getCourseId());
+        assertEquals(Const.StatusCodes.BACKDOOR_STATUS_SUCCESS, status);
+        verifyAbsentInDatastore(feedbackSession1InCourse1);
+
+        // verify if question of session is deleted
+        FeedbackQuestionAttributes question1InSession1InCourse1 =
+                dataBundle.feedbackQuestions.get("qn1InSession1InCourse1");
+        verifyAbsentInDatastore(question1InSession1InCourse1);
```
Comment (Crphang) [inline src/test/java/teammates/test/cases/testdriver/BackDoorTest.java]: As discussed, the cascading is supposed to occur. I think in order to test this, can we do a `verifyPresentInDatastore(question1InSession1InCourse1)`, before we do the deletion of the session?

Comment (tshradheya) [reply src/test/java/teammates/test/cases/testdriver/BackDoorTest.java]: Sure. Will add it

**Pull request merged**